src/render/painter.js

```diff
@@ -257,72 +266,166 @@ class Painter {
             if (sourceCache.used) {
                 sourceCache.prepare(this.gl);
             }
         }
 
-        this.showOverdrawInspector(options.showOverdrawInspector);
+        const layerIds = this.style._order;
 
-        this.depthRange = (style._order.length + 2) * this.numSublayers * this.depthEpsilon;
+        // 3D pass
+        // We first create a renderbuffer that we'll use to preserve depth
+        // results across 3D layers, then render each 3D layer to its own
+        // framebuffer/texture, which we'll use later in the translucent pass
+        // to render to the main framebuffer. By doing this before we render to
+        // the main framebuffer we won't have to do an expensive framebuffer
+        // restore mid-render pass.
+        // The most important distinction of the 3D pass is that we use the
+        // depth buffer in an entirely different way (to represent 3D space)
+        // than we do in the 2D pass (to preserve layer order).
+        this.renderPass = '3d';
+        {
+            const gl = this.gl;
+            // We'll wait and only attach the depth renderbuffer if we think we're
+            // rendering something.
+            let first = true;
+
+            let sourceCache;
+            let coords = [];
+
+            for (let i = 0; i < layerIds.length; i++) {
+                const layer = this.style._layers[layerIds[i]];
+
+                if (!layer.has3DPass() || layer.isHidden(this.transform.zoom)) continue;
+
+                if (layer.source !== (sourceCache && sourceCache.id)) {
+                    sourceCache = this.style.sourceCaches[layer.source];
+                    coords = [];
+
+                    if (sourceCache) {
+                        this.clearStencil();
+                        coords = sourceCache.getVisibleCoordinates();
+                    }
 
-        this.isOpaquePass = true;
-        this.renderPass();
-        this.isOpaquePass = false;
-        this.renderPass();
+                    coords.reverse();
+                }
 
-        if (this.options.showTileBoundaries) {
-            const sourceCache = this.style.sourceCaches[Object.keys(this.style.sourceCaches)[0]];
-            if (sourceCache) {
-                draw.debug(this, sourceCache, sourceCache.getVisibleCoordinates());
+                if (!coords.length) continue;
+
+                this._setup3DRenderbuffer();
+
+                const renderTarget = this.viewportFrames.pop() || new RenderTexture(this);
+                renderTarget.attachDepthRenderbuffer(this.depthRbo);
+
+                if (first) {
+                    this.clearDepth();
+                    first = false;
+                }
+
+                this.renderLayer(this, (sourceCache: any), layer, coords);
+
+                renderTarget.detachDepthRenderbuffer();
+                this.prerenderedFrames[layer.id] = renderTarget;
             }
+
+            if (this.depthRboAttached) gl.bindRenderbuffer(gl.RENDERBUFFER, null);
```

Comment:
`_setup3DRenderbuffer` always unbinds the renderbuffer, so is this necessary?

Reply:
Given that we're binding/unbinding framebuffers, I believe we shouldn't be unbinding renderbuffers manually, since this state is part of the FBO.